feat: add SDK support for search records with similarity

jfcalvo opened the pull request:
# Description

This PR suggest an implementation to give support to `/api/me/datasets/{dataset_id}/records/search` endpoint with vector similarity searching.

Here is a full example working with similarity search from python SDK using no real 100% fake data:
```python
rg.init(api_key=owner.api_key)
workspace = Workspace.create(name="test")

feedback_dataset.add_vector_settings(VectorSettings(name="vector", dimensions=2))

feedback_dataset.add_records(
    [
        FeedbackRecord(fields={"text": "hello"}, vectors={"vector": [1, 1]}),
        FeedbackRecord(
            fields={"text": "hello"},
            vectors={"vector": [2, 2]},
            responses=[ResponseSchema(status="discarded")],
        ),
        FeedbackRecord(
            fields={"text": "hello"},
            vectors={"vector": [4, 4]},
            responses=[ResponseSchema(status="submitted", values={"question": {"value": "answer"}})],
        ),
    ]
)

remote = feedback_dataset.push_to_argilla("test_find_similar_records", workspace=workspace)
only_submitted_and_discarded_records = remote.filter_by(response_status=["submitted", "discarded"])

records_with_scores = only_submitted_and_discarded_records.find_similar_records(
    vector_name="vector",
    value=[1, 1],
    max_results=3,
)
# Even if the used vector is the same that contains the record 0, the search will skip it since it has no submitted or discarded responses.
```

Closes #4020

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

Tested locally

**Checklist**

- [ ] I added relevant documentation
- [ ] follows the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [x] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

jfcalvo committed:
feat: add functionality draft to add new similarity search records

jfcalvo committed:
feat: add correct URL to search_records

jfcalvo committed:
feat: add some missing logic

frascuchon commented on src/argilla/client/sdk/v1/datasets/api.py:
```diff
@@ -215,6 +217,43 @@ def get_records(
+    offset: int = 0,
+    limit: int = 50,
+) -> Response[Union[FeedbackRecordsSearchModel, ErrorMessage, HTTPValidationError]]:
+    url = f"/api/me/datasets/{id}/records/search"
```
Once we have the "admin" endpoint, we should change by `/api/v1/datasets/{id}/records/search`
```suggestion
    url = f"/api/v1/me/datasets/{id}/records/search"
```

frascuchon commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -480,6 +480,10 @@ def add_records(
         self._records.add(records=records, show_progress=show_progress)
 
+    def search_records(self):
+        # TODO: Call here to argilla.client.sdk.v1.datasets.api.search_records ???
```
Yes, and we need to return a list of tuples of [`RemoteFeedbackRecord`, `float`(the schore)]

jfcalvo committed:
Merge branch 'feature/feedback-dataset-semantic-similarity' into feature/add-sdk-search-records

jfcalvo committed:
feat: add some improvements and tests

jfcalvo committed:
Merge branch 'feature/feedback-dataset-semantic-similarity' into feature/add-sdk-search-records

jfcalvo committed:
feat: align implementation with current product design

jfcalvo commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -480,6 +481,46 @@ def add_records(
+        self,
+        vector_name: str,
+        value: Optional[List[float]] = None,
+        record: Optional[RemoteFeedbackRecord] = None,
```
@alvarobartt is this signature ok? When we expect a record from the outside world it should be a `RemoteFeedbackRecord`?

alvarobartt replied:
No, users will never instantiate a `RemoteFeedbackRecord` but a `FeedbackRecord`, if we are planning to allow both then we should use the `Union`

jfcalvo replied:
The problem that I see is that `FeedbackRecord` doesn't has an `id` and we need the `id` to be used as identifier to search by that record.

frascuchon replied:
Since we need the `record_id` to find similar records for a given record, we need records already stored in Argilla, so `RemoteFeedbackRecord` objects. 

This method cannot work with records without an ID. The workflow to work with records in this method could be:
```python
argilla_ds = FeedbackDataset.from_argilla("dataset")

reference_record = argilla_ds[0]

simlar_records = argilla_ds.search_by(vector_name="my_vector", record=reference_record)

```

jfcalvo commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -480,6 +481,46 @@ def add_records(
+        value: Optional[List[float]] = None,
+        record: Optional[RemoteFeedbackRecord] = None,
+        max_results: int = 50,
+    ) -> List[Tuple[RemoteFeedbackRecord, float]]:
```
@alvarobartt the same here. It's ok to return `RemoteFeedbackRecord` here?

alvarobartt replied:
So the `RemoteFeedbackRecord` should just be returned if so, in this case since the search is applied in Argilla, then this is correct!

jfcalvo commented on tests/unit/client/feedback/dataset/remote/test_dataset.py:
```diff
@@ -117,6 +117,9 @@ def create_mock_routes(
 
 
 class TestSuiteRemoteDataset:
+    def test_search_records(self) -> None:
```
@alvarobartt apart from this missing test do you think there is another one that I need to implement?

alvarobartt reviewed (commented):
As already mentioned here I think we should use a client schema indeed, not the SDK one

alvarobartt reviewed (commented):
WDYT about using `pydantic` schemas in the SDK instead of just the kwargs? @frascuchon @gabrielmbmb

alvarobartt reviewed (commented):
WDYT about using `pydantic` schemas in the SDK instead of kwargs? @frascuchon @gabrielmbmb

frascuchon commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -480,6 +481,46 @@ def add_records(
+        Returns:
+            A list of tuples with each tuple including a record and a similarity score.
+        """
+        try:
```
We should allow to combine search with filters

frascuchon committed:
Merge branch 'feature/feedback-dataset-semantic-similarity' into feature/add-sdk-search-records

frascuchon committed:
refactor: Define find_similar_records as an abstract method and provide an implementation for both local and remote

frascuchon committed:
refactor: clean API request models and prepare body code block

frascuchon committed:
chore: add missing method for test base class

frascuchon committed:
tests: Adding more tests

frascuchon committed:
chore: Review type hints mismatch

frascuchon committed:
chore: Update changelog

frascuchon commented on CHANGELOG.md:
```diff
@@ -39,6 +39,7 @@ These are the section headers that we use:
 - Added `add_vector_settings` method to local and remote `FeedbackDataset`. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
 - Added `vector_settings_by_name` method to local and remote `FeedbackDataset`. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
 - Added `ARGILLA_SEARCH_ENGINE` environment variable to configure the search engine to use. ([#4019](https://github.com/argilla-io/argilla/pull/4019))
+- Added `find_similar_records method to local and remote `FeedbackDataset`. ([#4023](https://github.com/argilla-io/argilla/pull/4023)) 
```
```suggestion
- Added `find_similar_records` method to local and remote `FeedbackDataset`. ([#4023](https://github.com/argilla-io/argilla/pull/4023)) 
```

frascuchon committed:
Apply suggestions from code review

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

gabrielmbmb commented on src/argilla/client/sdk/v1/datasets/api.py:
```diff
@@ -217,6 +219,56 @@ def get_records(
     return handle_response_error(response)
 
 
+# TODO: We must change endpoint URL to "/api/v1/me/datasets/{id}/records/search" once the endpoint is available.
```
to `/api/v1/datasets/{id}/records/search`?

frascuchon replied:
Yes. We need this endpoint to return all user reponses

jfcalvo commented on tests/integration/client/sdk/v1/test_datasets.py:
```diff
@@ -387,6 +392,32 @@ async def test_get_records(role: UserRole) -> None:
     assert FeedbackItemModel(**response.parsed.items[0].dict())
 
 
+@pytest.mark.skip(reason="unfinished")
```
I set this as skipped because it was unfinished, we should remove it.

frascuchon replied:
Removing this and create proper sdk unit tests

frascuchon committed:
chore: Review client Api error exceptions

frascuchon committed:
tests: create unit tests checking search_records calls

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
feat: Using cosine similarity (#4124)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR changes the `l2_norm` distance to the cosine similarity for
vector search. This change can improve results on similarity searches
and also for least similarity searches.
 
This [PR](https://github.com/argilla-io/argilla/pull/4023) must be
reviewed first

Closes https://github.com/argilla-io/argilla/issues/4123

**Type of change**

(Please delete options that are not relevant. Remember to title the PR
according to the type of change)

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing
functionality)
- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And
ideally, reference `tests`)

The base dataset has been used with boh ElasticSearch and OpenSearch to
verify this change.

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

alvarobartt commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -310,6 +311,38 @@ def _validate_metadata_names(self):
+        self,
+        vector_name: str,
+        value: Optional[List[float]],
+        record: Optional[RemoteFeedbackRecord],
```
```suggestion
        value: Optional[List[float]] = None,
        record: Optional[RemoteFeedbackRecord] = None,
```

alvarobartt commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -310,6 +311,38 @@ def _validate_metadata_names(self):
+        value: Optional[List[float]],
+        record: Optional[RemoteFeedbackRecord],
+        max_results: int = 50,
+    ):
```
```suggestion
    ) -> List[Tuple[RemoteFeedbackRecord, float]]:
```

alvarobartt commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -310,6 +311,38 @@ def _validate_metadata_names(self):
+        record: Optional[RemoteFeedbackRecord],
+        max_results: int = 50,
+    ):
+        try:
```
We should add a check to see whether at least one of `value` or `record` is provided and raise a `ValueError` if both are provided or if any is provided

gabrielmbmb replied:
We're doing that on the `root_validator` of `FeedbackRecordsSearchVectorQuery`

alvarobartt replied:
Yes I know, but do we want to keep that logic there instead of on the client side?

alvarobartt commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -482,6 +515,27 @@ def add_records(
+            vector_name: a vector name to use for searching by similarity.
+            value: an optional vector value to be used for searching by similarity.
+            record: an optional record to be used for searching by similarity.
+            max_results: the maximum number of results for the search.
```
```suggestion
            value: an optional vector value to be used for searching by similarity. Defaults to None.
            record: an optional record to be used for searching by similarity. Defaults to None.
            max_results: the maximum number of results for the search. Defaults to 50.
```

alvarobartt replied:
And I'd also add a note that one of `value` or `record` needs to be provided, and not both.

alvarobartt commented on src/argilla/client/sdk/v1/datasets/api.py:
```diff
@@ -217,6 +219,56 @@ def get_records(
+        response_status: the status of the responses to be retrieved.
+            Can either be `draft`, `missing`, `discarded`, or `submitted`. Defaults to None.
+        metadata_filters: the metadata filters to be applied to the records. Defaults to None.
+        limit: an optional value to limit the number of returned records by the search.
```
```suggestion
        limit: an optional value to limit the number of returned records by the search. Defaults to 50.
```

alvarobartt commented on src/argilla/client/sdk/v1/datasets/models.py:
```diff
@@ -93,6 +93,35 @@ class FeedbackRecordsModel(BaseModel):
+    record_id: Optional[UUID] = None
+    value: Optional[List[float]] = None
+
+    @root_validator
+    def check_required(cls, values: dict) -> dict:
+        """Check that either 'record_id' or 'value' is provided"""
+        record_id = values.get("record_id")
+        value = values.get("value")
+
+        if bool(record_id) == bool(value):
+            raise ValueError("Either 'record_id' or 'value' must be provided")
+
+        return values
```
With the changes mentioned above I think this validation is not required here

frascuchon replied:
Let's leave this validation here, since it is aligned with the same schema of the server side. If this validation changes in the future, we can change this code without having to change anything anywhere else

alvarobartt reviewed (commented):
Left some minor comments, I'll complete the review once those are addressed, nice job!

gabrielmbmb commented on tests/unit/client/sdk/models/v1/test_search_records.py:
Shouldn't be this file under `sdk/v1` instead of `sdk/models/v1`?

frascuchon replied:
DOne

frascuchon committed:
chore: move file

frascuchon committed:
Update src/argilla/client/feedback/dataset/remote/dataset.py

Co-authored-by: Alvaro Bartolome <alvaro@argilla.io>

frascuchon committed:
Update src/argilla/client/sdk/v1/datasets/api.py

Co-authored-by: Alvaro Bartolome <alvaro@argilla.io>

frascuchon committed:
Merge branch 'feature/feedback-dataset-semantic-similarity' into feature/add-sdk-search-records

frascuchon committed:
chore: Review signature

frascuchon committed:
Merge branch 'feature/add-sdk-search-records' of github.com:argilla-io/argilla into feature/add-sdk-search-records
